asserts/asserts.go

```diff
@@ -0,0 +1,285 @@
+}
+
+// Header returns the value of an header by name.
+func (ab *AssertionBase) Header(name string) string {
+	return ab.headers[name]
+}
+
+// Body returns the body of the assertion.
+func (ab *AssertionBase) Body() []byte {
+	return ab.body
+}
+
+// Signature returns the signed content and its signature already split and decoded.
+func (ab *AssertionBase) Signature() (content []byte, sigtype string, signature []byte) {
+	return ab.content, ab.sigtype, ab.sigpacket
+}
+
+// sanity check
+var _ Assertion = (*AssertionBase)(nil)
+
+var (
+	nlnl = []byte("\n\n")
+
+	// for basic sanity checking of header names
+	headerNameSanity = regexp.MustCompile("^[a-z][a-z0-9-]*[a-z0-9]$")
+)
+
+func parseHeaders(head []byte) (map[string]string, error) {
+	if !utf8.Valid(head) {
+		return nil, fmt.Errorf("header is not utf8")
+	}
+	headers := make(map[string]string)
+	for _, entry := range strings.Split(string(head), "\n") {
+		nameValueSplit := strings.Index(entry, ": ")
+		if nameValueSplit == -1 {
+			return nil, fmt.Errorf("header entry missing name value ': ' separation: %q", entry)
+		}
+		name := entry[:nameValueSplit]
+		if !headerNameSanity.MatchString(name) {
+			return nil, fmt.Errorf("invalid header name: %v", name)
+		}
+		headers[name] = entry[nameValueSplit+2:]
+	}
+	return headers, nil
+}
+
+// Decode parses a serialized assertion.
+//
+// The expected serialisation format looks like:
+//
+//   HEADER ("\n\n" BODY?)? "\n\n" SIGNATURE
+//
+// where:
+//
+//    HEADER is a set of header lines separated by "\n"
+//    BODY can be arbitrary,
+//    SIGNATURE is composed like:
+//      SIG-TYPE " " BASE64-SIG-PACKET
+//
+//    SIG-TYPE is the type/format expected for the base64
+//    encoded signature value, for now only "openpgp" is supported
+//
+// An header line looks like:
+//
+//   NAME ": " VALUE
+//
+// for sanity NAME is expected to match headerNameSanity regexp.
+//
+// The following headers are mandatory:
+//
+//   type
+//   authority-id (the signer id)
+//   revision (a positive int)
+//   body-length (int expected to be equal to the length of BODY)
+//
+func Decode(serializedAssertion []byte) (Assertion, error) {
+	contentSignatureSplit := bytes.LastIndex(serializedAssertion, nlnl)
+	if contentSignatureSplit == -1 {
+		return nil, fmt.Errorf("assertion content/signature separator not found")
+	}
+	content := serializedAssertion[:contentSignatureSplit]
+	signature := serializedAssertion[contentSignatureSplit+2:]
+
+	headersBodySplit := bytes.Index(content, nlnl)
+	var body, head []byte
+	if headersBodySplit == -1 {
+		head = content
+	} else {
+		body = content[headersBodySplit+2:]
+		if len(body) == 0 {
+			body = nil
+		}
+		head = content[:headersBodySplit]
+	}
+
+	headers, err := parseHeaders(head)
+	if err != nil {
+		return nil, fmt.Errorf("parsing assertion headers: %v", err)
+	}
+
+	if len(signature) == 0 {
+		return nil, fmt.Errorf("empty assertion signature")
+	}
+
+	sigtypeSigpacketSplit := bytes.IndexByte(signature, ' ')
```

Comment:
Given comments above, all the decoding from here on would be moved into the signature-checking logic.